File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int winstart = 0, maxLength = 0;
        unordered_map<char, int> repeated_chars;
        for (int winend = 0; winend < s.size(); winend++) {
            while (repeated_chars.count(s[winend])) {
                repeated_chars.erase(s[winstart]);
                winstart++;
            }
            maxLength = max(winend - winstart + 1, maxLength);
            repeated_chars[s[winend]] = winend;
        }
        return maxLength;
    }
};

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashMap<Character, Integer> map = new HashMap<>();
        int winstart = 0;
        int maxLen = 0;
        for (int i = 0; i < s.length(); i++) {
            while (map.containsKey(s.charAt(i))) {
                map.remove(s.charAt(winstart));
                winstart++;
            }
            maxLen = Math.max(maxLen, i - winstart + 1);
            map.put(s.charAt(i), 1);
        }
        return maxLen;
    }
}